File: EMU7800.D2D/EMU7800.D2D.Interop.WinRT_8.1WP/TextFormat.cpp
// © Mike Murphy

#include "pch.h"
#include "TextFormat.h"

using namespace EMU7800::D2D::Interop;

TextFormat::TextFormat(IDWriteFactory* pDWriteFactory, String^ fontFamilyName, int fontWeight, int fontStyle, int fontStretch, FLOAT fontSize) :
    m_hr(0)
{
    if (!pDWriteFactory)
    {
        m_hr = E_POINTER;
        return;
    }

    m_hr = pDWriteFactory->CreateTextFormat(
        fontFamilyName->Data(),
        NULL,
        (DWRITE_FONT_WEIGHT)fontWeight,
        (DWRITE_FONT_STYLE)fontStyle,
        (DWRITE_FONT_STRETCH)fontStretch,
        fontSize,
        L"",
        &m_pTextFormat);
}

int TextFormat::SetTextAlignment(DWriteTextAlignment textAlignment)
{
    if FAILED(m_hr)
        return m_hr;

    HRESULT hr = m_pTextFormat->SetTextAlignment((DWRITE_TEXT_ALIGNMENT)textAlignment);
    return hr;
}

int TextFormat::SetParagraphAlignment(DWriteParaAlignment paragraphAlignment)
{
    if FAILED(m_hr)
        return m_hr;

    HRESULT hr = m_pTextFormat->SetParagraphAlignment((DWRITE_PARAGRAPH_ALIGNMENT)paragraphAlignment);
    return hr;
}

File: src/win32.interop/TextFormat.cs
using System;

using static EMU7800.Win32.Interop.Direct2DNativeMethods;

namespace EMU7800.Win32.Interop
{
    public class TextFormat : IDisposable
    {
        public IntPtr TextFormatPtr { get; private set; }

        public int HR { get; private set; }

        public void SetTextAlignment(DWriteTextAlignment textAlignment)
            => Direct2D_SetTextAlignmentForTextFormat(TextFormatPtr, textAlignment);

        public void SetParagraphAlignment(DWriteParaAlignment paragraphAlignment)
            => Direct2D_SetParagraphAlignmentForTextFormat(TextFormatPtr, paragraphAlignment);

        #region IDispose Members

        public void Dispose()
        {
            Direct2D_ReleaseTextFormat(TextFormatPtr);
            TextFormatPtr = IntPtr.Zero;
            HR = 0;
            GC.SuppressFinalize(this);
        }

        #endregion

        #region Constructors

        public TextFormat(string fontFamilyName, float fontSize)
        {
            var ptr = IntPtr.Zero;
            HR = Direct2D_CreateTextFormat(fontFamilyName, fontSize, ref ptr);
            TextFormatPtr = ptr;
        }

        public TextFormat(string fontFamilyName, int fontWeight, int fontStyle, int fontStretch,float fontSize)
        {
            var ptr = IntPtr.Zero;
            HR = Direct2D_CreateTextFormatEx(fontFamilyName, fontWeight, fontStyle, fontStretch, fontSize, ref ptr);
            TextFormatPtr = ptr;
        }

        #endregion
    }
}
